src/goto-programs/remove_function_pointers.cpp

```diff
@@ -140,6 +162,317 @@ symbolt &remove_function_pointerst::new_tmp_symbol()
+
+  Inputs:
+   symbol_expr - The expression that is the value of the function pointer.
+
+ Outputs: Returns true if the expression is a symbol_exprt and it was able
+          to either find a explict function that this points to (in the case
+          where the symbol is a function) or all a set of functions that it
+          could be by following the valu the symbol represents. In either of
+          the true cases the out_functions will contain all the functions the
+          function pointer could point to.
+
+ Purpose: Resolve a symbol to its real value and recursively try and find
+          all posible functions that could be at that position in the array.
+
+\*******************************************************************/
+
+bool remove_function_pointerst::try_get_call_from_symbol(
+  const exprt &symbol_expr, functionst &out_functions)
+{
+  if(symbol_expr.id()==ID_symbol)
+  {
+    const symbolt &symbol=
+      symbol_table.lookup(symbol_expr.get(ID_identifier));
+    exprt looked_up_val=symbol.value;
+    exprt precise_expr;
+    if(try_get_precise_call(looked_up_val, precise_expr))
+    {
+      out_functions.push_back(precise_expr);
+    }
+    else if(try_get_from_address_of(looked_up_val, out_functions))
+    {
+      return true;
+    }
+    else if(try_get_call_from_index(looked_up_val, out_functions))
+    {
+      return true;
+    }
+    else if(try_get_call_from_symbol(looked_up_val, out_functions))
+    {
+      return true;
+    }
+    else
+    {
+      return false;
+    }
+  }
+  else
+  {
+    return false;
+  }
+}
+
+/*******************************************************************\
+
+Function: remove_function_pointerst::try_get_call_from_index
+
+  Inputs:
+   expr - The expression that is the value of the function pointer.
+
+ Outputs: Returns true if the expression is a index_exprt and it was able
+          to either find a explict function that this points to (in the case
+          where the index is constant) or all the functions within the array
+          that the index is indexing. In either of the true cases the
+          out_functions will contain all the functions the function pointer
+          could point to.
+
+ Purpose: Resolve the index of an array to find all posible functions
+          that could be at that position in the array.
+
+\*******************************************************************/
+
+bool remove_function_pointerst::try_get_call_from_index(
+  const exprt &expr, functionst &out_functions)
+{
+  if(expr.id()==ID_index)
+  {
+    index_exprt index_expr=to_index_expr(expr);
+
+    array_exprt array_expr;
+    if(try_get_array_from_index(index_expr, array_expr))
+    {
+      mp_integer value;
+      if(try_evaluate_index_value(index_expr.index(), value))
+      {
+        const exprt &func_expr=array_expr.operands()[integer2size_t(value)];
+        exprt precise_match;
+        if(try_get_precise_call(func_expr, precise_match))
```

Comment:
These no-body if statements aren't very clean. I shall have a think about a better way to structure this code.